### DISCUSSION TITLE
skynet websocket client 粘包问题

### DISCUSSION THREAD

**Discussion by vvwo** [cloudwu/skynet#1837] [General]
场景：
client通过server的http post消息中获取到server的ip,port信息，
client得到host内容后向server发起handshake。
server升级websocket协议后，立即向client发送第一条websocket请求。

现象：
skynet websocket clent运行在同一个环境中，server运行在不同环境中，
极少数环境中，出现第一条websocket请求body不完整的现象。

定位：
server运行在不同环境，101 Switching Protocols与第一条websocket请求间隔不同。
间隔非常短时，100us左右或更快时，epoll的2次read event，分别读到64,128字节。
在websocket client处理handshake中会pop出来64+128字节的内容。
即，skynet/lualib/http/internal.lua中M.request的local body = M.recvheader(read, tmpline, "")
而，128字节中包含了第一条websocket请求的部分内容。

临时解决方案：server发送第一条websocket请求前延时100ms，现象不再出现。

TODO：期望云风大大优化一下。

**Comment by cloudwu** [comment]
我不明白这是什么问题。

但你可以考虑使用其它 websocket 模块，或改进 skynet 附带的 websocket 模块。

**Comment by cloudwu** [comment]
我看了一下，https://github.com/cloudwu/skynet/blob/master/lualib/http/websocket.lua#L86  这里调用 recvheader 后，返回值 `header_body`  不应该丢弃，而应该放在后面 https://github.com/cloudwu/skynet/blob/master/lualib/http/websocket.lua#L276 的 `recv_buf` 里面？

@lvzixun

**Reply by cloudwu** [reply]
@zorocn 你可以尝试修改一下，如果改好，可以提个 PR 。

**Comment by vvwo** [comment]
描述的可能有点误解，
指的是client write_handshake时，
即https://github.com/cloudwu/skynet/blob/master/lualib/http/websocket.lua#L53
在https://github.com/cloudwu/skynet/blob/master/lualib/http/internal.lua#L191 里面的
https://github.com/cloudwu/skynet/blob/master/lualib/http/internal.lua#49 pop出来了64+128字节的内容

**Reply by cloudwu** [reply]
recvheader 这个函数设计的是：返回值是多余没有处理的内容，需要交到下一个环节处理。websocket 的握手过程直接丢弃了它的返回值，这是不对的。

**Reply by cloudwu** [reply]
建议的修改方法，加一个函数
```lua
local function reader_with_payload(self, payload)
	local sz_payload = #payload
	if sz_payload == 0 then
		return
	end
	local read = self.read
	function self.read (sz)
		if sz == nil or sz == sz_payload then
			self.read = read
			return payload
		end
		if sz < sz_payload then
			local ret = payload:sub(1, sz)
			payload = payload:sub(sz + 1)
			return ret
		end
		self.read = read
		return payload .. read(sz - sz_payload)
	end
end
```
然后在 https://github.com/cloudwu/skynet/blob/master/lualib/http/websocket.lua#L163 read_handshake 最后调用一下，把 self.read 改掉：

```lua
reader_with_payload(self,header_body)
```


上面代码没有测试过，有劳 @zorocn 试试看能不能修好这个 bug ，然后提个 pr 。

**Reply by lvzixun** [reply]
仔细看了下代码， 貌似只能这么改了。本来想改到 `tlshelper.lua` 里面，把`read_buff` 给加回去，但是对于非tls的ws是直接用`sockethelper.lua` 读取的，并没有`read_buff` 可以修改。 我这边提交下PR，@zorocn 测试下吧。

**Comment by vvwo** [comment]
> recvheader 这个函数设计的是：返回值是多余没有处理的内容，需要交到下一个环节处理。websocket 的握手过程直接丢弃了它的返回值，这是不对的。

原因正是这个。
由于版本比较旧，只是合入了websocket client相关部分，问题显现在write_handshake中。read_handshake的确有同样问题。

**Reply by lvzixun** [reply]
能否帮助测试下这个PR？
https://github.com/cloudwu/skynet/pull/1839 应该是能够fix你的问题。

**Reply by vvwo** [reply]
抱歉，我也想，由于版本非常旧，是个原始都没有websocket.lua文件的版本。这次合入和使用的client部分。
这个思路是否可行，若recvheader 有返回值是多余没有处理的内容( https://github.com/cloudwu/skynet/blob/master/lualib/http/websocket.lua#L53 和  https://github.com/cloudwu/skynet/blob/master/lualib/http/websocket.lua#L86 )，赋给self.payload。
另，初始化obj时， read = sockethelper.readfunc(socket_id, self.payload)。

**Reply by cloudwu** [reply]
所以说，你就没打算解决这个问题？那干嘛提 issue ?

**Reply by vvwo** [reply]
不不不，这边环境没有wss，ws显现问题也只用到了client。
在write_handshake中有返回多余没处理内容时，验证了reader_with_payload函数的思路可以解决。
read_handshake同理，无环境验证。
